File: app/src/main/java/com/kaiserpudding/howtheywrite/characterDetail/CharacterDetailActivity.java
package com.kaiserpudding.howtheywrite.characterDetail;

import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.TextView;
import com.kaiserpudding.howtheywrite.R;
import com.kaiserpudding.howtheywrite.model.Character;

public class CharacterDetailActivity extends AppCompatActivity {

  public static String REPLY_CHARACTER_ID = "howTheyWrite.CHARACTER_ID";

  private TextView kanjiTextView;
  private TextView hiraganaTextView;
  private TextView translationTextView;

  private CharacterDetailViewModel characterDetailViewModel;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_character_detail);

    //TODO proper handling for 0
    int characterId = getIntent().getIntExtra(REPLY_CHARACTER_ID, 0);

    //TODO proper fix for racing condition of initChar and getChar
    characterDetailViewModel = ViewModelProviders.of(
        this, new CharacterDetailViewModelFactory(getApplication(), characterId))
        .get(CharacterDetailViewModel.class);
    Character character = null ;

    while (character == null) {
      character = characterDetailViewModel.getCharacter();
    }

    kanjiTextView = findViewById(R.id.kanji);
    kanjiTextView.setText(character.getKanji());
    hiraganaTextView = findViewById(R.id.hiragana);
    hiraganaTextView.setText(character.getHiragana());
    translationTextView = findViewById(R.id.translation);
    if (character.getTranslationKey() != null) {
      translationTextView.setText(
          getResources().getIdentifier(
              character.getTranslationKey(),
              "string",
              getApplicationContext().getPackageName()));
    } else {
      translationTextView.setText(character.getTranslation());
    }
  }
}

File: HowTheyWrite_android/app/src/main/java/com/kaiserpudding/howtheywrite/characterDetail/CharacterDetailActivity.kt
//package com.kaiserpudding.howtheywrite.characterDetail
//
//import androidx.lifecycle.ViewModelProviders
//import androidx.appcompat.app.AppCompatActivity
//import android.os.Bundle
//import android.widget.TextView
//import com.kaiserpudding.howtheywrite.R
//import com.kaiserpudding.howtheywrite.model.Character
//
//class CharacterDetailActivity : AppCompatActivity() {
//
//    private var hanziTextView: TextView? = null
//    private var pinyinTextView: TextView? = null
//    private var translationTextView: TextView? = null
//
//    private var characterDetailViewModel: CharacterDetailViewModel? = null
//
//    override fun onCreate(savedInstanceState: Bundle?) {
//        super.onCreate(savedInstanceState)
//        setContentView(R.layout.activity_character_detail)
//
//        //TODO proper handling for 0
//        val characterId = intent.getIntExtra(REPLY_CHARACTER_ID, 0)
//
//        //TODO proper fix for racing condition of initChar and getChar
//        characterDetailViewModel = ViewModelProviders.of(
//                this, CharacterDetailViewModelFactory(application, characterId))
//                .get(CharacterDetailViewModel::class.java)
//        var character: Character? = null
//
//        while (character == null) {
//            character = characterDetailViewModel!!.character
//        }
//
//        hanziTextView = findViewById(R.id.hanzi)
//        hanziTextView!!.text = character.hanzi
//        pinyinTextView = findViewById(R.id.pinyin)
//        pinyinTextView!!.text = character.pinyin
//        translationTextView = findViewById(R.id.translation)
//        if (character.translationKey != null) {
//            translationTextView!!.setText(
//                    resources.getIdentifier(
//                            character.translationKey,
//                            "string",
//                            applicationContext.packageName
//                    )
//            )
//        } else {
//            translationTextView!!.text = character.translation
//        }
//    }
//
//    companion object {
//
//        const val REPLY_CHARACTER_ID = "howTheyWrite.CHARACTER_ID"
//    }
//}
